### ISSUE TITLE
k8s_raw ingress rules not handled idempotently

### ISSUE THREAD

**Issue by travisgroth** [ansible/ansible#38312]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

k8s_raw
##### ANSIBLE VERSION

```
ansible 2.5.0
  config file = /Users/tgroth/.ansible.cfg
  configured module search path = [u'/Users/tgroth/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/tgroth/Applications/ansible/lib/python2.7/site-packages/ansible
  executable location = /Users/tgroth/Applications/ansible/bin/ansible
  python version = 2.7.14 (default, Sep 22 2017, 00:05:22) [GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)]
```

##### CONFIGURATION

CACHE_PLUGIN(/Users/tgroth/.ansible.cfg) = jsonfile
CACHE_PLUGIN_CONNECTION(/Users/tgroth/.ansible.cfg) = /tmp/fact_cache
CACHE_PLUGIN_TIMEOUT(/Users/tgroth/.ansible.cfg) = 600
DEFAULT_HOST_LIST(env: ANSIBLE_INVENTORY) = [u'/Users/tgroth/.ansible/hosts']
DEFAULT_TRANSPORT(/Users/tgroth/.ansible.cfg) = ssh
##### OS / ENVIRONMENT

N/A
##### SUMMARY

When managing an ingress resource with host rules in the spec, k8s_raw does not handle the rules correctly.  Specifically, upon every task run, another copy of rules in your spec will be added, even if they are identical to those already on the ingress resource.  
##### STEPS TO REPRODUCE

```yaml
---
- hosts: localhost
  tasks:
    - k8s_raw:
        name: nginx
        definition:
          apiVersion: v1beta1
          kind: Ingress
          metadata:
            name: nginx
            namespace: default
          spec:
            rules:
            - host: vhost1
              http:
                paths:
                - backend:
                    serviceName: vhost1
                    servicePort: 80
            - host: vhost2
              http:
                paths:
                - backend:
                    serviceName: vhost2
                    servicePort: 80
            - host: vhost3
              http:
                paths:
                - backend:
                    serviceName: vhost3
                    servicePort: 80

```
Run this playbook **twice** against minikube and then see results from kubectl get ingress nginx.  

##### EXPECTED RESULTS

```
changed: [localhost] => {
    "changed": true,
    "invocation": {
        "module_args": {
            "api_key": null,
            "cert_file": null,
            "context": "minikube",
            "host": null,
            "key_file": null,
            "kubeconfig": null,
            "name": "nginx",
            "namespace": "default",
            "password": null,
            "spec_rules": [
                {
                    "host": "vhost1",
                    "http": {
                        "paths": [
                            {
                                "backend": {
                                    "serviceName": "vhost1",
                                    "servicePort": 80
                                }
                            }
                        ]
                    }
                },
                {
                    "host": "vhost2",
                    "http": {
                        "paths": [
                            {
                                "backend": {
                                    "serviceName": "vhost2",
                                    "servicePort": 80
                                }
                            }
                        ]
                    }
                },
                {
                    "host": "vhost3",
                    "http": {
                        "paths": [
                            {
                                "backend": {
                                    "serviceName": "vhost3",
                                    "servicePort": 80
                                }
                            }
                        ]
                    }
                }
            ],
            "ssl_ca_cert": null,
            "username": null,
            "verify_ssl": null
        }
    },
    "result": {
        "api_version": "extensions/v1beta1",
        "kind": "Ingress",
        "metadata": {
            "annotations": null,
            "cluster_name": null,
            "creation_timestamp": "2018-04-05T02:08:23+00:00",
            "deletion_grace_period_seconds": null,
            "deletion_timestamp": null,
            "finalizers": null,
            "generate_name": null,
            "generation": 1,
            "initializers": null,
            "labels": null,
            "name": "nginx",
            "namespace": "default",
            "owner_references": null,
            "resource_version": "795042",
            "self_link": "/apis/extensions/v1beta1/namespaces/default/ingresses/nginx",
            "uid": "37ecda44-3876-11e8-814d-0800273d4951"
        },
        "spec": {
            "backend": null,
            "rules": [
                {
                    "host": "vhost1",
                    "http": {
                        "paths": [
                            {
                                "backend": {
                                    "service_name": "vhost1",
                                    "service_port": 80
                                },
                                "path": null
                            }
                        ]
                    }
                },
                {
                    "host": "vhost2",
                    "http": {
                        "paths": [
                            {
                                "backend": {
                                    "service_name": "vhost2",
                                    "service_port": 80
                                },
                                "path": null
                            }
                        ]
                    }
                },
                {
                    "host": "vhost3",
                    "http": {
                        "paths": [
                            {
                                "backend": {
                                    "service_name": "vhost3",
                                    "service_port": 80
                                },
                                "path": null
                            }
                        ]
                    }
                }
            ],
            "tls": null
        },
        "status": {
            "load_balancer": {
                "ingress": null
            }
        }
    }
}

$ kubectl get ingress nginx -o yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  creationTimestamp: 2018-04-05T02:08:23Z
  generation: 2
  name: nginx
  namespace: default
  resourceVersion: "795110"
  selfLink: /apis/extensions/v1beta1/namespaces/default/ingresses/nginx
  uid: 37ecda44-3876-11e8-814d-0800273d4951
spec:
  rules:
  - host: vhost1
    http:
      paths:
      - backend:
          serviceName: vhost1
          servicePort: 80
  - host: vhost2
    http:
      paths:
      - backend:
          serviceName: vhost2
          servicePort: 80
  - host: vhost3
    http:
      paths:
      - backend:
          serviceName: vhost3
          servicePort: 80
status:
  loadBalancer:
    ingress:
    - ip: 192.168.99.101
```
##### ACTUAL RESULTS

```
changed: [localhost] => {
    "changed": true,
    "invocation": {
        "module_args": {
            "api_key": null,
            "cert_file": null,
            "context": "minikube",
            "host": null,
            "key_file": null,
            "kubeconfig": null,
            "name": "nginx",
            "namespace": "default",
            "password": null,
            "spec_rules": [
                {
                    "host": "vhost1",
                    "http": {
                        "paths": [
                            {
                                "backend": {
                                    "serviceName": "vhost1",
                                    "servicePort": 80
                                }
                            }
                        ]
                    }
                },
                {
                    "host": "vhost2",
                    "http": {
                        "paths": [
                            {
                                "backend": {
                                    "serviceName": "vhost2",
                                    "servicePort": 80
                                }
                            }
                        ]
                    }
                },
                {
                    "host": "vhost3",
                    "http": {
                        "paths": [
                            {
                                "backend": {
                                    "serviceName": "vhost3",
                                    "servicePort": 80
                                }
                            }
                        ]
                    }
                }
            ],
            "ssl_ca_cert": null,
            "username": null,
            "verify_ssl": null
        }
    },
    "result": {
        "api_version": "extensions/v1beta1",
        "kind": "Ingress",
        "metadata": {
            "annotations": null,
            "cluster_name": null,
            "creation_timestamp": "2018-04-05T02:08:23+00:00",
            "deletion_grace_period_seconds": null,
            "deletion_timestamp": null,
            "finalizers": null,
            "generate_name": null,
            "generation": 2,
            "initializers": null,
            "labels": null,
            "name": "nginx",
            "namespace": "default",
            "owner_references": null,
            "resource_version": "795110",
            "self_link": "/apis/extensions/v1beta1/namespaces/default/ingresses/nginx",
            "uid": "37ecda44-3876-11e8-814d-0800273d4951"
        },
        "spec": {
            "backend": null,
            "rules": [
                {
                    "host": "vhost1",
                    "http": {
                        "paths": [
                            {
                                "backend": {
                                    "service_name": "vhost1",
                                    "service_port": 80
                                },
                                "path": null
                            }
                        ]
                    }
                },
                {
                    "host": "vhost2",
                    "http": {
                        "paths": [
                            {
                                "backend": {
                                    "service_name": "vhost2",
                                    "service_port": 80
                                },
                                "path": null
                            }
                        ]
                    }
                },
                {
                    "host": "vhost3",
                    "http": {
                        "paths": [
                            {
                                "backend": {
                                    "service_name": "vhost3",
                                    "service_port": 80
                                },
                                "path": null
                            }
                        ]
                    }
                },
                {
                    "host": "vhost1",
                    "http": {
                        "paths": [
                            {
                                "backend": {
                                    "service_name": "vhost1",
                                    "service_port": 80
                                },
                                "path": null
                            }
                        ]
                    }
                },
                {
                    "host": "vhost2",
                    "http": {
                        "paths": [
                            {
                                "backend": {
                                    "service_name": "vhost2",
                                    "service_port": 80
                                },
                                "path": null
                            }
                        ]
                    }
                },
                {
                    "host": "vhost3",
                    "http": {
                        "paths": [
                            {
                                "backend": {
                                    "service_name": "vhost3",
                                    "service_port": 80
                                },
                                "path": null
                            }
                        ]
                    }
                }
            ],
            "tls": null
        },
        "status": {
            "load_balancer": {
                "ingress": [
                    {
                        "hostname": null,
                        "ip": "192.168.99.101"
                    }
                ]
            }
        }
    }
}

$ kubectl get ingress nginx -o yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  creationTimestamp: 2018-04-05T02:08:23Z
  generation: 2
  name: nginx
  namespace: default
  resourceVersion: "795110"
  selfLink: /apis/extensions/v1beta1/namespaces/default/ingresses/nginx
  uid: 37ecda44-3876-11e8-814d-0800273d4951
spec:
  rules:
  - host: vhost1
    http:
      paths:
      - backend:
          serviceName: vhost1
          servicePort: 80
  - host: vhost2
    http:
      paths:
      - backend:
          serviceName: vhost2
          servicePort: 80
  - host: vhost3
    http:
      paths:
      - backend:
          serviceName: vhost3
          servicePort: 80
  - host: vhost1
    http:
      paths:
      - backend:
          serviceName: vhost1
          servicePort: 80
  - host: vhost2
    http:
      paths:
      - backend:
          serviceName: vhost2
          servicePort: 80
  - host: vhost3
    http:
      paths:
      - backend:
          serviceName: vhost3
          servicePort: 80
status:
  loadBalancer:
    ingress:
    - ip: 192.168.99.101
```

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/modules/clustering/k8s/k8s_raw.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/clustering/k8s/k8s_raw.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @chouseknecht @fabianvf @flaper87 @maxamillion @sdoran
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by travisgroth** [comment]
Any feedback on this?  Ingress controllers are the only piece of our typical k8s deployments that we can't put into ansible right now.

**Comment by tv** [comment]
This is blocking my work as well. I had time today to debug the issue further and it looks like list inside of lists are not diffed recursively [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/k8s/helper.py#L474). It's not Ingress specific, but lists inside of lists without name keys are quite rare in the kubernetes api.
I'll try to fix this and make PR tomorrow.

**Comment by travisgroth** [comment]
@tv ah that make sense.  Awesome!  Happy to try the PR out on our end.

**Comment by tv** [comment]
Bad news. This was much bigger job than I had anticipated. It wasn't really an one day job to get this done correctly. And I had to get back to "real" work, so I reverted back to just using kubectl with shell module, which I was using before this module landed in 2.5.0.

It's just simple wrapper which writes kubeconfig from template variables and uses it with `kubectl --kubeconfig {{ kubeconfig.path }} apply -f -` with stdin being the yaml to be submitted. It's simple, and it's genuinely raw, but it works until this is fixed.

Sorry to say, that I don't think I have time to fix this in near future.
